setOnBackgroundMessage() is now a Future<void> that will await for the function to complete.

asoap opened the pull request:
This is in regards to this issue here:
https://github.com/ably/ably-flutter/issues/496

If a setOnBackgroundMessage() was fired and if it took to long to complete the isolate on Android would end before the code was completed.  This was extra annoying if you were trying to debug a background messsage or if it used other packages that used async/await.  Those items just wouldn't fire. 

Now all code will complete before the isolate dismisses itself.

asoap commented:
Do I need to go and fix these errors that have popped up?  There is an AWS error for credentials, I'm not sure how I'm supposed to get past that for a github test.

ttypic commented on lib/src/push_notifications/src/push_notification_events.dart:
```diff
@@ -4,7 +4,7 @@ import 'package:ably_flutter/ably_flutter.dart';
 /// Type definition for a callback invoked when background push message
 /// is received.
-typedef BackgroundMessageHandler = void Function(RemoteMessage message);
+typedef BackgroundMessageHandler = Future<void> Function(RemoteMessage message);
```
It seems this change introduces breaking change, can we avoid this? We are not ready right now to create next major release of Flutter SDK

asoap replied:
I think this was needed to make the async await for it.  I'll play around with it.  I think we can change the type to be "dynamic" instead of void/Future<void>.  That might satisfy both situations.  

Is the issue it creates this one? 

```dart
 info • `Future` results in `async` function bodies must be `await`ed or marked `unawaited` using `dart:async` • lib/src/platform/src/method_call_handler.dart:97:5 • unawaited_futures
```

ttypic replied:
Awesome! Thank you

ttypic reviewed (changes requested):
Thank you very much for this pull request. We cannot merge it right now because it appears to introduce breaking changes. However, we are happy to work on this together.

ttypic commented:
> Do I need to go and fix these errors that have popped up? There is an AWS error for credentials, I'm not sure how I'm supposed to get past that for a github test.

No need, this is because actions run on forked repository and it doesn't have access to secrets

ttypic reviewed (changes requested):
Thank you very much for updated version, looks good! But could you please run `flutter format .` there are some extra whitespaces and linter doesn't like it. Also I am not sure we need to do any changes in `lib/src/platform/src/background_android_isolate_platform.dart `

asoap commented on lib/src/platform/src/background_android_isolate_platform.dart:
```diff
@@ -20,7 +20,8 @@ class BackgroundIsolateAndroidPlatform {
       switch (call.method) {
         case PlatformMethod.pushOnBackgroundMessage:
-          return _onPushBackgroundMessage(call.arguments as RemoteMessage);
+          await _onPushBackgroundMessage(call.arguments as RemoteMessage);
```
That one I'm fairly sure it needs to be await.  That's the code for the isolate.  When that code ends it ends the isolate.  I believe if it returns it will just return Future<void>, then closes the isolate without fully executing the _onPushBackgroundMessage().  

My first test for this change was to add a delay here, which worked, but wasn't an ideal solution.

ttypic replied:
@asoap yes, but this `async` function and it has some syntax sugar, e.g.  returned `Future` should be unwrapped (basically awaited) automatically, otherwise it would return `Future<Future<void>>`.

sacOO7 replied:
Can we do it like 
```
      dynamic onBackgroundMessageResult = _onPushBackgroundMessage(call.arguments as RemoteMessage);
      // ^^^ _onBackgroundMessage() can return void or Future<void>
      if (onBackgroundMessageResult is Future) {        
        await onBackgroundMessageResult; // if it returns a Future we await on it.
      }
```
If you are not 100% sure if its a future

sacOO7 replied:
@ttypic AFAIK, if we have nested promises, generally all of them resolve in one go with `await`. Not sure how it works in flutter though

ttypic replied:
@sacOO7 Futures are flatten here for sure, this code works right now :) and `_onPushBackgroundMessage ` returns `Future<void>` right now, there were no changes, I think you are confused with  `onPushBackgroundMessage` from `lib/src/platform/src/push_notification_events_internal.dart`

asoap replied:
@ttypic  The _onPushBackgroundMessage() call inside of background_android_isolate_platform.dart calls handleBackgroundMessage() in push_notification_events_internal.dart   That's why I changed the handleBackgroundMessage() to be a Future<void> as well. This is setup in a way where these two functions are doing the same thing.  They both check to see if there is a function to call, and then call it.   The _onPushBackgroundMessage() should always get a Future<void> from handleBackgroundMessage();

asoap commented:
> Thank you very much for updated version, looks good! But could you please run `flutter format .` there are some extra whitespaces and linter doesn't like it. Also I am not sure we need to do any changes in `lib/src/platform/src/background_android_isolate_platform.dart `

I added a comment above.  I'm fairly sure that change is required.  I can test without it if you like to confirm. 

Also I'm a bit confused on how to run flutter format.  Obviously the command line, but for these changes I made them locally in my installed packages for flutter, and I made a fork and changed them online.  I guess I could pull my repo locally, run the command and then push it.   But I'm not sure if that would work.  Working on packages is all new to me.

asoap commented:
I pushed the changes to my repo.  Also I was confused looking for the command "flutter format ." It turns out it was "dart format ."

I think everything should be good to go now.

asoap commented:
Huzzah!  Is this done now?  Or do I need to do anything to resolve the conversation?

ttypic commented:
@asoap Sorry, still some checks [are not passing](https://github.com/ably/ably-flutter/actions/runs/7118880398/job/19382767695?pr=498), so I can't merge it, could you add those changes:

`lib/src/platform/src/push_notification_events_internal.dart`

<img width="1299" alt="image" src="https://github.com/ably/ably-flutter/assets/16488867/047a698c-94b6-46d0-97b5-d9483c87aa0d">

`lib/src/platform/src/method_call_handler.dart`

<img width="1287" alt="image" src="https://github.com/ably/ably-flutter/assets/16488867/31c9b75d-66be-4bc2-b325-84ba5c62dc17">

asoap commented:
Can do.  I'm going to be in the middle of some work, but I will try to address this today and test it out to make sure it works before submitting changes.

asoap commented:
Unfortuantely work went long today, and I'll have to address this on Monday.

asoap commented:
I tested these changes and everything is working as expected.  I then pushed them to the pull request.

ttypic reviewed (changes requested):
Hello @asoap I'm sorry but formatting checks are still failing and I can't merge. Do you mind if I close the PR and add required changes in new internal PR?

asoap commented:
> Hello @asoap I'm sorry but formatting checks are still failing and I can't merge. Do you mind if I close the PR and add required changes in new internal PR?

Yeah absolutely, do what you need to do.

ttypic commented:
moved to https://github.com/ably/ably-flutter/pull/499
